File: MRU/MRUFileList.vb
Imports System.Windows.Controls
Imports System.Windows

Public Class MRUFileList
    Inherits MenuItem

#Region "Classes"
    Public Class FileSelectedEventArgs
        Inherits EventArgs

        Public Sub New(ByVal fileName As String)
            _fileName = fileName
        End Sub

        Private _fileName As String

        Public ReadOnly Property FileName As String
            Get
                Return _fileName
            End Get
        End Property
    End Class
#End Region

#Region "Constructor"
    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(ByVal storage As IStorage)
        MyBase.New()
        Me.Storage = storage
    End Sub
#End Region

#Region "Events"
    Public Event FileSelected As EventHandler(Of FileSelectedEventArgs)
    Public Event MaxItemsChanged As EventHandler
#End Region

#Region "Members"
    Private _maximumItems As Int32 = 10
#End Region

    Public Shared StorageProperty As DependencyProperty = DependencyProperty.Register("Storage",
                                                                                      GetType(IStorage),
                                                                                      GetType(MRUFileList),
                                                                                      New PropertyMetadata(Nothing,
                                                                                                           AddressOf StoragePropertyChanged))

#Region "Public Properties"
    Public Property Storage As IStorage
        Get
            Return DirectCast(GetValue(StorageProperty), IStorage)
        End Get
        Set(value As IStorage)
            SetValue(StorageProperty, value)
        End Set
    End Property

    Public Property MaximumItems As Int32
        Get
            Return _maximumItems
        End Get
        Set(value As Int32)
            If _maximumItems <> value Then
                _maximumItems = value
            End If
        End Set
    End Property
#End Region

#Region "Public Methods"
    Public Sub AddItem(ByVal filename As String)
        Dim item As MRUFileListItem = GetItem(filename)
        If item IsNot Nothing Then Items.Remove(item)
        item = New MRUFileListItem(filename)
        Items.Insert(0, item)
        AddHandler item.Click, AddressOf mnuItem_Click
        UpdateStorage()
    End Sub

    Public Sub RemoveItem(ByVal filename As String)
        Dim item As MRUFileListItem = GetItem(filename)
        If item IsNot Nothing Then
            RemoveHandler item.Click, AddressOf mnuItem_Click
            Items.Remove(item)
        End If
        UpdateStorage()
    End Sub

    Public Sub RemoveItem(ByVal index As Int32)
        Dim item As MRUFileListItem = CType(Items(index), MRUFileListItem)
        RemoveHandler item.Click, AddressOf mnuItem_Click
        Items.Remove(item)
        UpdateStorage()
    End Sub

    Public Sub RemoveItem(ByVal item As MRUFileListItem)
        RemoveHandler item.Click, AddressOf mnuItem_Click
        Items.Remove(item)
        UpdateStorage()
    End Sub

    Public Function GetItem(ByVal filename As String) As MRUFileListItem
        Dim index As Int32 = -1
        For i As Int32 = 0 To Items.Count - 1
            If DirectCast(Me.Items(i), MRUFileListItem).FileName.Equals(filename, StringComparison.CurrentCultureIgnoreCase) Then
                index = i
            End If
        Next
        If index >= 0 AndAlso index < Items.Count Then Return DirectCast(Items(index), MRUFileListItem)
        Return Nothing
    End Function
#End Region

#Region "Protected Methods"
    Protected Overridable Sub UpdateStorage()
        Dim fileList As New List(Of String)
        For Each item As MRUFileListItem In Items
            fileList.Add(item.FileName)
        Next
        Storage.WriteFileList(fileList)
    End Sub

    Protected Overridable Sub OnFileSelected(ByVal fileName As String)
        RaiseEvent FileSelected(Me, New FileSelectedEventArgs(fileName))
    End Sub

    Protected Overridable Sub OnMaxItemsChanged()
        RaiseEvent MaxItemsChanged(Me, EventArgs.Empty)
        Do While Me.Items.Count > MaximumItems
            Me.Items.RemoveAt(Me.Items.Count - 1)
        Loop
    End Sub
#End Region

#Region "Event Handlers"
    Private Shared Sub StoragePropertyChanged(d As DependencyObject, e As DependencyPropertyChangedEventArgs)
        Dim instance As MRUFileList = DirectCast(d, MRUFileList)
        Dim fileList As IEnumerable(Of String) = instance.Storage.ReadFileList
        instance.Items.Clear()
        For Each fileName As String In fileList
            instance.AddItem(fileName)
        Next
    End Sub

    Private Sub mnuItem_Click(sender As Object, e As RoutedEventArgs)
        RaiseEvent FileSelected(sender, New FileSelectedEventArgs(DirectCast(sender, MRUFileListItem).FileName))
    End Sub
#End Region

End Class

File: MRU/MRUFileList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Controls;
using System.Windows;

namespace MRU
{
    public class MRUFileList : MenuItem
    {
        
        #region Classes
        public class FileSelectedEventArgs : EventArgs
        {
            private string _fileName;
            
            public FileSelectedEventArgs(string fileName)
            {
                _fileName = fileName;
            }

            public string FileName { get { return _fileName; } }
        }
        #endregion

        #region Constructors
        public MRUFileList(): base()
        {            
        }

        public MRUFileList(IStorage storage): base()
        {
            this.Storage = storage;
        }

        #endregion

        #region Events
        public event EventHandler<FileSelectedEventArgs> FileSelected;
        public event EventHandler MaxItemsChanged;
        #endregion

        #region Dependency Properties
        public static DependencyProperty StorageProperty = DependencyProperty.Register("Storage",
                                                                                       typeof(IStorage),
                                                                                       typeof(MRUFileList),
                                                                                       new PropertyMetadata(null, StoragePropertyChanged));
        #endregion

        #region Members
        private Int32 _maximumItems = 10;
        #endregion

        #region Properties
        public IStorage Storage { get; set; }
        public Int32 MaximumItems
        {
            get { return _maximumItems; }
            set 
            { 
                if (_maximumItems != value)
                {
                    _maximumItems = value;
                }
            }
        }
        #endregion
        
        #region Methods
        public void AddItem(string fileName)
        {
            MRUFileListItem item = GetItem(fileName);
            if (item != null) { Items.Remove(item); }
            item = new MRUFileListItem(fileName);
            Items.Insert(0, item);

            item.Click += mnuItem_Clicked;
            UpdateStorage();
        }

        public MRUFileListItem GetItem(string fileName)
        {
            Int32 index = -1;
            for (Int32 i = 0; i < Items.Count; i++)
            {
                MRUFileListItem current = (MRUFileListItem)Items[i];
                if (current.FileName.Equals(fileName, StringComparison.CurrentCultureIgnoreCase))
                {
                    index = i;
                }
            }
            if (index >= 0 && index < Items.Count)
            {
                return (MRUFileListItem)Items[index];
            }
            return null;
        }

        public void RemoveItem(string fileName)
        {
            MRUFileListItem item = GetItem(fileName);
            if (item != null)
            {
                item.Click -= mnuItem_Clicked;
                Items.Remove(item);
            }
            UpdateStorage();
        }

        public void RemoveItem(Int32 index)
        {
            MRUFileListItem item = (MRUFileListItem)Items[index];
            item.Click -= mnuItem_Clicked;
            Items.Remove(item);
            UpdateStorage();
        }

        public void RemoveItem(MRUFileListItem item)
        {
            item.Click -= mnuItem_Clicked;
            Items.Remove(item);
            UpdateStorage();
        }
        
        protected virtual void OnFileSelected(string fileName)
        {
            FileSelected(this, new FileSelectedEventArgs(fileName));
        }

        protected virtual void OnMaxItemsChanged()
        {
            MaxItemsChanged(this, EventArgs.Empty);
            while (this.Items.Count > MaximumItems)
            {
                Items.RemoveAt(Items.Count - 1);
            }
        }

        protected virtual void UpdateStorage()
        {
            List<string> fileList = new List<string>();
            foreach (MRUFileListItem item in Items)
            {
                fileList.Add(item.FileName);
            }
            Storage.WriteFileList(fileList);
        }
        #endregion  

        #region Event Handlers
        private static void StoragePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs  e)
        {
            MRUFileList instance = ((object)d) as MRUFileList;
            IEnumerable<string> fileList = instance.Storage.ReadFileList();
            instance.Items.Clear();
            foreach (string fileName in fileList)
            {
                instance.AddItem(fileName);
            }

        }

        private void mnuItem_Clicked(object sender, RoutedEventArgs e)
        {
            MRUFileListItem instance = (MRUFileListItem)sender;
            OnFileSelected(instance.FileName);
        }
        #endregion
        


    }
}
